src/pushsource/_impl/backend/pub_source/pub_source.py

```diff
@@ -0,0 +1,106 @@
+import logging
+from concurrent import futures
+
+from more_executors import Executors
+
+from ...helpers import (
+    as_completed_with_timeout_reset,
+    force_https,
+    list_argument,
+    try_int,
+)
+from ...model import AmiPushItem
+from ...source import Source
+from .pub_client import PubClient
+
+LOG = logging.getLogger("pushsource")
+
+
+class PubSource(Source):
+    """Uses push item list from Pub as the source of push items."""
+
+    def __init__(
+        self,
+        url,
+        task_id,
+        threads=4,
+        timeout=60 * 60 * 4,
+    ):
+        """Create a new source.
+
+        Parameters:
+            url (src)
+                Base URL of Pub Tool, e.g. "http://pub.example.com",
+                "https://pub.example.com:8123".
+
+            task_id (int, str, list[str], list[int])
+                Task ID(s) to be used as push item source.
+                If a single string is given, multiple IDs may be
+                comma-separated.
+
+            threads (int)
+                Number of threads used for concurrent queries to Pub.
+
+            timeout (int)
+                Number of seconds after which an error is raised, if no progress is
+                made during queries to Pub.
+        """
+
+        self._url = force_https(url)
+        self._task_ids = self._validate_task_ids(
+            [try_int(x) for x in list_argument(task_id)]
+        )
+        self._client = PubClient(threads=threads, url=self._url)
+        self._timeout = timeout
+        # This executor doesn't use retry because pub-client executor already does that.
+        self._executor = Executors.thread_pool(
+            name="pushsource-pub", max_workers=threads
+        ).with_cancel_on_shutdown()
+
+    def _validate_task_ids(self, task_ids):
+        out = []
+        for task_id in task_ids:
+            if isinstance(task_id, int):
+                out.append(task_id)
+            else:
+                LOG.warning("Invalid task_id: %s", task_id)
```

Comment:
This should probably mention Pub somewhere.  In a complex task with a lot of log messages it seems like it might not be obvious this is referring to a Pub task ID.